modules/apps/cookies/cookies-impl/src/main/java/com/liferay/cookies/internal/events/CookiesPreAction.java

```diff
@@ -0,0 +1,274 @@
+		}
+	}
+
+	private void _expireOptionalCookies(
+		HttpServletRequest httpServletRequest,
+		HttpServletResponse httpServletResponse) {
+
+		for (String optionalCookieName : CookiesPreAction._OPTIONAL_COOKIES) {
+			_expireCookie(
+				httpServletRequest, httpServletResponse, optionalCookieName);
+		}
+	}
+
+	private void _expireUserConsentConfiguredCookie(
+		HttpServletRequest httpServletRequest,
+		HttpServletResponse httpServletResponse) {
+
+		_expireCookie(
+			httpServletRequest, httpServletResponse,
+			CookiesConstants.NAME_USER_CONSENT_CONFIGURED);
+	}
+
+	private Map<String, String> _parseCookieMap(Cookie[] cookies) {
+		Map<String, String> cookieMap = new HashMap<>();
+
+		if (cookies != null) {
+			for (Cookie cookie : cookies) {
+				String cookieName = cookie.getName();
+
+				if (cookieName.startsWith("CONSENT_TYPE_") ||
+					cookieName.equals(
+						CookiesConstants.NAME_USER_CONSENT_CONFIGURED)) {
+
+					cookieMap.put(cookieName, cookie.getValue());
+				}
+			}
+		}
+
+		return cookieMap;
+	}
+
+	private void _run(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws ConfigurationException {
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		CookiesPreferenceHandlingConfiguration
+			cookiesPreferenceHandlingConfiguration =
+				_configurationProvider.getGroupConfiguration(
+					CookiesPreferenceHandlingConfiguration.class,
+					themeDisplay.getScopeGroupId());
+
+		Map<String, String> requestCookies = _parseCookieMap(
+			httpServletRequest.getCookies());
+
+		String necessaryConsentCookie = requestCookies.get(
+			CookiesConstants.NAME_CONSENT_TYPE_NECESSARY);
+		String performanceConsentCookie = requestCookies.get(
+			CookiesConstants.NAME_CONSENT_TYPE_PERFORMANCE);
+		String functionalConsentCookie = requestCookies.get(
+			CookiesConstants.NAME_CONSENT_TYPE_FUNCTIONAL);
+		String personalizationConsentCookie = requestCookies.get(
+			CookiesConstants.NAME_CONSENT_TYPE_PERSONALIZATION);
+		String userConsentCookie = requestCookies.get(
+			CookiesConstants.NAME_USER_CONSENT_CONFIGURED);
+
+		boolean optionalCookiesSet = false;
+
+		if ((performanceConsentCookie != null) &&
+			(functionalConsentCookie != null) &&
+			(personalizationConsentCookie != null)) {
+
+			optionalCookiesSet = true;
+		}
+
+		boolean allConsentCookiesSet = false;
+
+		if ((necessaryConsentCookie != null) && optionalCookiesSet) {
+			allConsentCookiesSet = true;
+		}
+
+		if (!cookiesPreferenceHandlingConfiguration.enabled()) {
+			if (userConsentCookie != null) {
+				_expireUserConsentConfiguredCookie(
+					httpServletRequest, httpServletResponse);
+			}
+
+			if (necessaryConsentCookie != null) {
+				_expireNecessaryCookies(
+					httpServletRequest, httpServletResponse);
+			}
+
+			if (optionalCookiesSet) {
+				_expireOptionalCookies(httpServletRequest, httpServletResponse);
+			}
+		}
+		else {
```

Comment:
All these conditions are a little messy but I guess is because of the cookie management logic and I don't know how to make it a little bit better. Maybe extracting the code I've mentioned in the comment above it improves a little.

Don't know if we can combine some condition checkings to improve readability

Reply:
I tried changing the order of conditions to simplify the code. I think it's a little less readable (it's harder to tell apart what will happen in explicit vs implicit mode), but it avoids repeated code